### DISCUSSION TITLE
Is MAUI mature enough for enterprise grade native application deployment?

### DISCUSSION THREAD

**Discussion by PBTests** [dotnet/maui#9478] [Polls]
Is MAUI mature enough for enterprise grade native application deployment?

**Comment by Adelzu** [comment]
As of September 2022, the short answer is "it depends"
Almost everything is fixable. But it requires time. IF and only IF you are not on tight deadlines, the answer is not yet.

**Comment by janseris** [comment]
Yes it is possible we are doing it but it will get your devs insane because almost every feature has a bug or a workaround is needed to make it working which is a lot of unwanted complexity in the project
It takes 4x more time to develop than it should

**Comment by Adelzu** [comment]
I changed my mind; With MAUI, you spend more time debugging the tools than debugging the app.
I wish we didn't move from Xamarin yet.

**Reply by manishbisen** [reply]
Sad to hear that

**Comment by GeoSaffer** [comment]
Its now June 2023......

Moving from Xamarin Forms to Maui feels like it is at least 2 years too early, but still Microsoft is forcing us to move to Maui. This is absolute bullshit ! It feels like the start of Xamarin forms All over again, FFS Microsoft wake up !

**Comment by Adelzu** [comment]
@GeoSaffer thanks for reminding me to post here again.

Here we are 8 months later. Now you :

1. debug the tools less, face weird issues.
2. trace weird issues and find out they known bugs and at of the github issue, you see a comment from the infamous ```msftbot``` saying ```We've added this issue to our backlog, and we will work to address it as time and resources allow``` or that the fix for a major issue is coming in the next major release!
3. try to find workarounds and fail around 50% of the time.
4. abandon the change then design it in another way.

It seems we are hitting a sunk cost fallacy with MAUI.

**Comment by nor0x** [comment]
> I changed my mind; With MAUI, you spend more time debugging the tools than debugging the app.
> I wish we didn't move from Xamarin yet.

unfortunately this is still true today, in the current state MAUI development is a great pita!

**Comment by FM1973** [comment]
It kind of depends....

We have a relatively big app in open beta for Android and Ios. So far, there are not many problems.
But we had to work aroud a lot of known bugs, as most of you. We are currently stuck at version 7.0.59 beause of the severe regression bugs in 7.0.81 and 7.0.86.

The main problems, in my oppinion, are the regression bugs and the bug fix speed for .net 7. You can´t update without having to test everything again, in my case, two platforms - if there is a version you can update to.

The second problem is the development policy. The team currently is targeting .net 8, as @hartez stated in https://github.com/dotnet/maui/issues/14257#issuecomment-1588195990 . I´d really like to knwo how the team decides which bug fixes are important enough to be backported to .net 7, since everyone of us has different needs. I think there are a lot, but the last service release for .net 7 was more than a month ago and there are a lot of important bug fixes we need to see in .net 7. Furtheremore this policy won´t change. So, even if we get a .net 8 version in november (!!!), the same game will start again. The team will develop against a beta version of .net 9 we can´t use, and we will be waiting for bugfixes to be backported to .net 8. Sisyphus is calling.....

So I can´t vote. I´m in between....

**Reply by Adelzu** [reply]
Your comment ended a 2 days spiral of weird issues. We have mistakenly upgraded from 7.0.59 and were face issues we couldn't explain!
Thank you.

**Comment by AlleSchonWeg** [comment]
Coupling to  .Net releases was the worst decision they have made. 
Nobody knows what bugs are fixed. In a service release or in next main version?! Totally random.

**Comment by giuseppenovielli** [comment]
@FM1973 @davidortinau 
With .NET 8, MAUI will be released also like nuget package, as Xamarin.Forms did?
So with nuget package can release bugs fix, fastest than .NET Versioning?

**Reply by FM1973** [reply]
Look there: [https://devblogs.microsoft.com/dotnet/announcing-dotnet-maui-in-dotnet-8-preview-3/](https://devblogs.microsoft.com/dotnet/announcing-dotnet-maui-in-dotnet-8-preview-3/) .

**Reply by giuseppenovielli** [reply]
Guys, the idea is that must provide a channel (like nuget) where can release fastly new MAUI version, for regression or some issues with hight priority.
So also 2 service release in a week, but only the most important issues, without waiting months, for fix an issue regression, that have generated another Service Release. 

P.S. This is only an'idea, i don't known how .NET Servicing work, but it can be very useful fix fastly regression, hight priority issues, that have generated by another Service Release.

**Comment by Strypper** [comment]
It still a total crap today with RC2

**Reply by Adelzu** [reply]
I think it is less buggy now. Still a long way to go though.

**Reply by Strypper** [reply]
Collectionview view still broken 
Label isn't display html correctly 
And people migrate crash the application with 0 logs on Android 
LOL this is a piece of garbage.

**Reply by azrinsani** [reply]
Is collection views still broken?

**Comment by GeoSaffer** [comment]
Still dont have the basics working, still stupid bugs and missing ported code. FFS Microsoft.

React is starting to look really good right now.

What a waist of time and money trying to get everything ported to Maui

**Comment by patkozlowski** [comment]
Unfortunatley, I had to migrate away after spending a few weeks porting apps over.  I cannot wait around for constant bug fixes for some very trivial things such as a basic performant/lag free collectionview.  The ball as been dropped leaving us Forms devs to be left out in the cold.  The idea is great but not so the execution.  I wanted this to work, but it just doesn't.  Memory leaks galore with basic components was my breaking point when apps are more complex than a todo app.  This is unacceptable on a customer facing app.  The workarounds created today will require workarounds or break in the future.  I said to myself "what is the point then to continue" if I have to put workaround upon workarounds, it is not ready for any production customer facing app.  I know it will be stable in the future but I cannot wait around for it.  I moved to react-native using expo.  It has been a great dev experience, I once feel the warmth again when things just work out of the box.

**Reply by JoeVoo** [reply]
React-native is no alternative to maui. It is Android and IOS only (as i see in a short look) and it is META. They closed 25000 issues, but look how they do. 
This poll is too old that it could represent the actual state. A lot of things are better now.

**Reply by patkozlowski** [reply]
Yeah I upgraded to .net 9 still most of the same - some regressions as well.  I have very complex layouts that just eat up memory.  Other than a very basic POC app, complex views and layouts will just not work due to internal memory leaks.  Forms was way better, it had its quirks but it did work.  My experience, fresh app with minimal functionality works well.  Once you add more complex stuff, memory expands and the OS just kills the app.  I believe every complex app will just error out if you are not internally doing many workarounds to release memory (of which should be automatically done by the controls and the views itself).  There is no "best practice" to handle this - ideally we shouldn't have to deal with such memory issues.  It will be ironed out in the future, but I just can't wait around.  Had to switch to another platform.  Will revist again in a few years when things are alot more stable as I do want this to succeed.

**Reply by true-perfect-code** [reply]
@patkozlowski 

Which framework are you switching to?

I spent a whole day a few days ago looking for alternatives to MAUI due to problems with the local Windows notification (which is supposed to be core functionality of MAUI).

After a long search I couldn't find anything that met my requirements except MAUI, so I'm still investing my private time in the hope that this year will be better.

My requirements for the framework are:
- UI should be possible with HTML/CSS and program logic (e.g. like Razor)
- one code base for all platforms (mobile, Windows and web) => this is possible with MAUI Blazor and Blazow Web
- Windows is very important to me, because if you want to create larger and more complex apps, you can't do that on a mobile display.

I tried Flutter more than a year ago and successfully created a project, but I really struggle with widgets (I just can't get to grips with them). 

Everything else I have found (e.g. Expo) does not work on mobile but Windows or vice versa (e.g. Elektron NET). Although React is not native HTML/CSS but it reminds me a bit of this approach.

So I'd be really interested to know if there are alternatives to MAUI so that I can finally make good use of my free time and don't have to keep looking for MAUI workarounds.

Thanks
pc

_PS: To be fair, you have to admit that MAUI itself is not always to fault, but also everything around it that does not meet the requirements of a modern app (specifically here WindowsAppSDK and no scheduling of the notification for MAUI, or visual Studio and publication with MAU Blazor A RCL + Blazow Web etc.)._

**Reply by patkozlowski** [reply]
Hi @true-perfect-code,

Yes, react native is more a web based structure to render stuff natively on mobile.  I use Typescript so it's pretty easy to wrap your head around.  I'm using it with expo, and it does support web as well.  You have a heavy desktop requirement so maybe its not a whole solution for you (as I don't require a desktop app).  You may want to try out Avalonia UI as I've heard good things and it's c# based so you can get up and running quick, but I think the web is in its infantcy and the ecosystem is a little smaller.  I haven't used it personally, but I've heard good things.  While react native is not a true "desktop app," you can package your web app as a desktop application using frameworks like Electron or Tauri, probably not the ideal.  My suggestion look at Avalonia UI and make sure you research all the requirements on your app before realizing something doesn't work.  You can try a hybrid approach as well as I've seen the best platforms focus more on desktop or mobile, but not the jack of all trades with everything.

**Reply by true-perfect-code** [reply]
@patkozlowski 

Thanks for the explanation. In the meantime, my feelings have calmed down because a way has been found to implement the MAUI Windows local notification via UWP (workaround).

So I'm sticking with MAUI because I currently don't see any alternative that would allow me to develop professionally for desktop Windows under the conditions that I have with MAUI Blazor (Razor/HTML/CSS). I've also been with NET since it was introduced (25 years now).

I really hope that things will get better.

pc

**Reply by patkozlowski** [reply]
It will, I have confidence in the team as they are taking on a massive task to get everything stable with the limited resources they have - by far very talented people.  It will get better and better in due time.  Just the growing pains of new tech.  Glad you figured it out!

**Reply by marcojak** [reply]
The concern is not about the team, but as you pointed out, the limited resources they have.

**Reply by MartyIX** [reply]
https://github.com/dotnet/maui/graphs/contributors does not look too bad.

**Reply by janseris** [reply]
The result is clear, regardless of graphs, that the framework is still buggy, 2 and half years after its release. Basic API extensions with very high vote count are not implemented and bugs are not fixed because there are too few people who have too few time. There are old bugs carried from Xamarin.Forms and the underlying Windows platform which are neither fixed. The amount of testers for bugs (bug verificators) is not properly allocated because they test bugs months or even 1+ years after they have been added to issues.
